### PR TITLE
[Collapsible] functional component and no jank when opening

### PR DESCRIPTION
**Pull request by alex-page**
### WHY are these changes introduced?

Fixes https://github.com/Shopify/polaris-react/issues/3390

### WHAT is this pull request doing?

- [x] Convert collapsible to a functional component
- [x] Fix janky movement when opening collapsible filters
- [x] Remove opacity from animation
- [x] Support nested collapsible elements

### How to 🎩

1. Go to the collapsible example
2. Go to the filters example and open and close the filters

### 🎩 checklist

* [x] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [x] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [x] Updated the component's `README.md` with documentation changes
* [x] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
* [x] For visual design changes, pinged one of @ HYPD, @ mirualves, @ sarahill, or @ ry5n to update the Polaris UI kit

### CODE REVIEW

File: src/components/Collapsible/tests/Collapsible.test.tsx [hunk 1]
```diff
@@ -17,22 +17,6 @@ describe('<Collapsible />', () => {
...
-    const collapsible = mountWithAppProvider(
-      <Collapsible id="test-collapsible" open>
-        <Child />
-      </Collapsible>,
-    );
-
-    expect(collapsible.find(Child)).toHaveLength(1);
-    collapsible.setProps({open: false});
-    collapsible.simulate('transitionEnd');
-    expect(collapsible.find(Child)).toHaveLength(0);
```
Comment (alex-page) [inline src/components/Collapsible/tests/Collapsible.test.tsx:35]: I couldn't see any great benefit to removing and adding back the content. So I am removing the tests that would test for this.

**Comment by BPScott** [comment]
Yay hookification. Not reviewed this but could you update https://github.com/Shopify/polaris-react/issues/1995 when it gets merged?

File: src/components/Collapsible/Collapsible.scss [hunk 2]
```diff
@@ -1,31 +1,27 @@
...
+.open:not(.animating) {
+  overflow: visible;
 }
 
-.fullyOpen {
-  overflow: visible;
+// Stop children from being focused when aria-hidden
+.Collapsible[aria-hidden='true']:not(.animating) {
+  display: none;
 }
```
Comment (alex-page) [inline src/components/Collapsible/Collapsible.scss:21]: Stop children from being focusable when aria-hidden

File: src/components/Collapsible/Collapsible.tsx [hunk 3]
```diff
@@ -30,165 +24,58 @@ export interface CollapsibleProps {
...
+    expandOnPrint && styles.expandOnPrint,
+    open && styles.open,
+    isAnimating && styles.animating,
+  );
+
+  useEffect(() => {
+    // isAnimating is true when the state changes not on page load
+    if (open !== isOpen) {
+      setIsAnimating(true);
+      setIsOpen(open);
```
Comment (alex-page) [inline src/components/Collapsible/Collapsible.tsx:57]: isAnimating is true when the state changes not on page load

**Comment by kyledurand** [comment]
I'm seeing a bit of animation jank when closing

[Before this pr](https://screenshot.click/screencast_2020-11-11_09-23-22.mp4)

[This pr](https://screenshot.click/screencast_2020-11-11_09-22-52.mp4)

File: src/components/Collapsible/Collapsible.tsx [hunk 4]
```diff
@@ -30,165 +24,58 @@ export interface CollapsibleProps {
...
-  | 'openingStart'
-  | 'opening';
-
-interface State {
-  height?: number | null;
-  animationState: AnimationState;
-  open: boolean;
-}
-
-const ParentCollapsibleExpandingContext = createContext(false);
```
Comment (AndrewMusgrave) [inline src/components/Collapsible/Collapsible.tsx:47]: This is needed for nested collapsibles
```
import React, {useState, useCallback} from 'react';

import {Page, Card, Stack, Button, Collapsible, TextContainer} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      <CollapsibleExample />
    </Page>
  );
}

function CollapsibleExample() {
  const [open, setOpen] = useState(true);
  const [innerOpen, setInnerOpen] = useState(true);

  const handleToggle = useCallback(() => setOpen((open) => !open), []);
  const handleInnerToggle = useCallback(
    () => setInnerOpen((open) => !open),
    [],
  );

  return (
    <div style={{height: '200px'}}>
      <Card sectioned>
        <Stack vertical>
          <Button
            onClick={handleToggle}
            ariaExpanded={open}
            ariaControls="basic-collapsible"
          >
            Toggle
          </Button>
          <Button
            onClick={handleInnerToggle}
            ariaExpanded={innerOpen}
            ariaControls="basic-collapsible"
          >
            Toggle inner
          </Button>
          <Collapsible
            open={open}
            id="basic-collapsible"
            transition={{duration: '500ms', timingFunction: 'ease'}}
            expandOnPrint
          >
            <TextContainer>
              Your mailing list lets you contact customers or visitors who have
              shown an interest in your store. Reach out to them with exclusive
              offers or updates about your products.
            </TextContainer>
            <Collapsible
              open={innerOpen}
              id="basic-inner-collapsible"
              transition={{duration: '500ms', timingFunction: 'ease-out'}}
              expandOnPrint
            >
              <TextContainer>
                Your mailing list lets you contact customers or visitors who
                have shown an interest in your store. Reach out to them with
                exclusive offers or updates about your products.
              </TextContainer>
            </Collapsible>
          </Collapsible>
        </Stack>
      </Card>
    </div>
  );
}
```

Comment (alex-page) [reply src/components/Collapsible/Collapsible.tsx:47]: @AndrewMusgrave this works with this PR, am I missing something?

File: src/components/Collapsible/Collapsible.tsx [hunk 5]
```diff
@@ -30,165 +24,78 @@ export interface CollapsibleProps {
...
-    height: null,
-    animationState: 'idle',
-    // eslint-disable-next-line react/no-unused-state
-    open: this.props.open,
+// Open
+// Add the height to trigger transition from 0 -> content height
+// Complete the transition and change the open state to add the class name and clean up inline styles
+
+// Close
+// Add the starting height
```
Comment (alex-page) [inline src/components/Collapsible/Collapsible.tsx:32]:
```suggestion
```

File: src/components/Collapsible/Collapsible.tsx [hunk 6]
```diff
@@ -30,165 +24,68 @@ export interface CollapsibleProps {
...
-        case 'openingStart':
-          this.setState({
-            animationState: 'opening',
-            height: heightNode ? heightNode.scrollHeight : 0,
-          });
-      }
-    });
-  }
+    setHeight(collapisbleContainer.current.scrollHeight);
+  }, [open]);
```
Comment (alex-page) [inline src/components/Collapsible/Collapsible.tsx:63]: This useEffect measured the component and applies the height. For open this is the only effect ran as it measures and applies, this causes the transition from 0 -> measured height. For closing we do the same thing but then use the second useEffect to transition to 0.

File: src/components/Collapsible/Collapsible.tsx [hunk 7]
```diff
@@ -30,165 +24,68 @@ export interface CollapsibleProps {
...
 
-function collapsibleHeight(
-  open: boolean,
-  animationState: AnimationState,
-  height?: number | null,
-) {
-  if (animationState === 'idle' && open) {
-    return open ? 'none' : undefined;
-  }
+    getComputedStyle(collapisbleContainer.current).height;
```
Comment (alex-page) [inline src/components/Collapsible/Collapsible.tsx:70]: This is needed otherwise the rendering engine measures and applies the height then set the new height causing no transition. This causes the rendering engine to apply the 0 height on the next cycle causing the transition to run.

File: src/components/Collapsible/README.md [hunk 8]
```diff
@@ -90,6 +91,70 @@ function CollapsibleExample() {
 }
 ```
 
+### Nested collapsible
```
Comment (alex-page) [inline src/components/Collapsible/README.md:94]: Is there a way to exclude examples from the styleguide? This isn't best practice but is important that it works.

**Comment by alex-page** [comment]
@AndrewMusgrave @kyledurand this should be good now.

File: src/components/Collapsible/Collapsible.scss [hunk 9]
```diff
@@ -1,33 +1,30 @@
...
+  height: auto;
+  overflow: visible;
 }
 
-.fullyOpen {
-  overflow: visible;
+// Stop children from being focused when aria-hidden
+.Collapsible[aria-hidden='true'] {
+  display: none;
 }
```
Comment (alex-page) [inline src/components/Collapsible/Collapsible.scss:22]: Stop closed content from being accessible from keyboard navigation.

**Comment by kyledurand** [comment]
Maybe the smallest regression. Not sure if it's a blocker https://screenshot.click/screencast_2020-11-13_16-02-48.mp4

File: src/components/Collapsible/Collapsible.scss [hunk 10]
```diff
@@ -1,33 +1,30 @@
...
-}
-
-.animating {
-  transition-property: opacity, max-height;
-  transition-duration: duration(slow);
+  height: 0;
+  overflow: hidden;
+  will-change: height;
+  transition-property: height;
+  transition-duration: duration(fast);
```
Comment (kyledurand) [inline src/components/Collapsible/Collapsible.scss:10]: Do we want to use `var(--p-duration-1-0-0)` here instead? 

Edit: nvm looks like we need to use tokens to match some JS below

File: src/components/Collapsible/README.md (hunk 8 above)
Comment (kyledurand) [reply src/components/Collapsible/README.md:94]: We'd need something like `<!-- example-for: none -->` but I don't think we have one at the moment

**Review by kyledurand — APPROVED** [review]
Code looks good 👍

**Comment by alex-page** [comment]
Would we ever want it to finish the animation before closing? I think you would want it to respond instantly on click. I could be wrong though..

File: src/components/Collapsible/README.md (hunk 8 above)
Comment (AndrewMusgrave) [reply src/components/Collapsible/README.md:94]: Would be great if we could grab context from the admin where nested collapsible are used to provide context on valid use cases. Unfortunately, I can't remember where it's being used, only that is it 😕

File: src/components/Collapsible/README.md [hunk 11]
```diff
@@ -90,6 +91,70 @@ function CollapsibleExample() {
...
+
+When you have multiple collapsibles inside each other. This should be avoided as it causes the user to open multiple collapsible areas.
+
+```jsx
+function CollapsibleExample() {
```
Comment (AndrewMusgrave) [inline src/components/Collapsible/README.md:99]:
```suggestion
function NestedCollapsibleExample() {
```

File: src/components/Collapsible/tests/Collapsible.test.tsx [hunk 12]
```diff
@@ -17,22 +17,6 @@ describe('<Collapsible />', () => {
 
     const hidden = collapsible.find(ariaHiddenSelector);
     expect(hidden.exists()).toBe(true);
-    expect(collapsible.contains('content')).toBe(false);
-  });
-
-  it('does not render its children when going from open to closed', () => {
```
Comment (AndrewMusgrave) [inline src/components/Collapsible/tests/Collapsible.test.tsx:23]: Unfortunately, this is an important test. Some consumer implementations rely on this - I'm sure you're going to see some tests failing in web

File: src/components/Collapsible/Collapsible.scss [hunk 13]
```diff
@@ -1,33 +1,30 @@
 @import '../../styles/common';
 
 .Collapsible {
-  overflow: hidden;
-  max-height: 0;
```
Comment (AndrewMusgrave) [inline src/components/Collapsible/Collapsible.scss:5]: Just writing a note to myself to tophat this: https://github.com/Shopify/polaris-react/pull/1980

Comment (AndrewMusgrave) [reply src/components/Collapsible/Collapsible.scss:5]: Looks good 😍

**Review by AndrewMusgrave — APPROVED** [review]
Screen Recording 2020-11-20 at 11 46 32 AM

The animation was acting oddly when opening/closing the collapsible quickly. Sometimes it didn't do anything, other times the animation would jank. It's an edge case so, whatever you think is best. Keep in mind the animation is at 10% speed.

Left a note about rendering children. Some tests in web expect on that - not sure about their implementations though. Just something to keep in mind when releasing this.

**Comment by johanstromqvist** [comment]
Looks great!! Thanks for the storyboard link @alex-page.

How is duration determined? If the value is fixed, a 500px collapse will be perceived as 10x the speed of a 50px collapse.

Dynamic containers benefit from dynamically determined durations. @AndrewMusgrave and I collaborated once to create a dynamic interpolation with normalized speed and clamped min/max values. Could that perhaps be applied?

File: src/components/Collapsible/tests/Collapsible.test.tsx (hunk 12 above)
Comment (alex-page) [reply src/components/Collapsible/tests/Collapsible.test.tsx:23]: Is it necessary to not render or tests will now fail?

**Comment by alex-page** [comment]
@johanstromqvist made a new issue for this https://github.com/Shopify/polaris-react/issues/3682

**Comment by johanstromqvist** [comment]
The interruptibility seems wonky when the transition is triggered a second time before it has finished. Only on collapse.
toggle wonk

File: src/components/Collapsible/tests/Collapsible.test.tsx (hunk 12 above)
Comment (AndrewMusgrave) [reply src/components/Collapsible/tests/Collapsible.test.tsx:23]: Tests will probably fail (unless the code has been refactored 🤞 ) - it might be easier to release a rc and run CI with that branch. I don't *think* UI will break - but it would be worthwhile checking the failed tests.

File: src/components/Collapsible/README.md (hunk 8 above)
Comment (BPScott) [reply src/components/Collapsible/README.md:94]: Add `../src/components/**/*.stories.tsx` to https://github.com/Shopify/polaris-react/blob/master/.storybook/main.js#L18

Then create a Collapsible.stories.tsx and dump your storybook stories in there using [the standard CSF format](https://storybook.js.org/docs/react/api/csf)

File: src/components/Collapsible/tests/Collapsible.test.tsx (hunk 12 above)
Comment (BPScott) [reply src/components/Collapsible/tests/Collapsible.test.tsx:23]: I'd like to remind y'all about https://buildkite.com/shopify/polaris-react-system-integration-test which runs a pass of the web CI on arbitary polaris branches.

It runs by default on our master branch, but it'd be trivial to trigger a manual build that points at this branch.

No faffing with RCs needed.
